src/lib.rs

```diff
@@ -72,6 +72,15 @@ pub trait Region : Default {
         region.reserve_regions(regions);
         region
     }
+
+    /// Determine this region's memory used and reserved capacity in bytes.
+    ///
+    /// An implementation should invoke the `callback` for each distinct allocation, providing the
+    /// used capacity as the first parameter and the actual capacity as the second parameter.
+    /// Both parameters represent a number of bytes.
+    /// The closure is free to sum the parameters, or do more advanced analysis such as creating a
+    /// histogram of allocation sizes.
+    fn heap_size(&self, callback: impl FnMut(usize, usize));
```

Comment:
You might ask why doesn't this return an `Iterator`? Well, this would mean that we'd need an associated type as each implementation returns a different iterator. Would work, but seems clumsy. I checked the assembly produced here, and the callbacks are all inlined, so I'm not worried about a performance overhead.